jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/ConcurrentPoolTest.java

```diff
@@ -611,4 +610,133 @@ public void testRandomStrategy()
         assertThat(e3.getPooled().get(), greaterThan(10));
         assertThat(e4.getPooled().get(), greaterThan(10));
     }
+
+    @ParameterizedTest
+    @MethodSource(value = "factories")
+    public void testLeakReserved(Factory factory)
+    {
+        ConcurrentPool<String> pool = factory.newPool(3);
+        Pool.Entry<String> e0 = pool.reserve();
+        Pool.Entry<String> e1 = pool.reserve();
+        Pool.Entry<String> e2 = pool.reserve();
+        assertThat(e0, notNullValue());
+        assertThat(e1, notNullValue());
+        assertThat(e2, notNullValue());
+
+        assertThat(pool.reserve(), nullValue());
+        assertThat(pool.size(), is(3));
+        assertThat(pool.getReservedCount(), is(3));
+        assertThat(pool.getIdleCount(), is(0));
+        assertThat(pool.getInUseCount(), is(0));
+
+        e0 = null;
+        System.gc();
+        pool.sweep();
+
+        assertThat(pool.size(), is(2));
+        assertThat(pool.getReservedCount(), is(2));
+        assertThat(pool.getIdleCount(), is(0));
+        assertThat(pool.getInUseCount(), is(0));
+
+        assertThat(e0, nullValue());
+    }
+
+    @ParameterizedTest
+    @MethodSource(value = "factories")
+    public void testLeakEnabled(Factory factory)
+    {
+        ConcurrentPool<String> pool = factory.newPool(10);
+        pool.reserve().enable("idle", false);
+        assertThat(pool.size(), is(1));
+        assertThat(pool.getReservedCount(), is(0));
+        assertThat(pool.getIdleCount(), is(1));
+        assertThat(pool.getInUseCount(), is(0));
+
+        Pool.Entry<String> e1 = pool.reserve();
+        e1.enable("held", true);
+        assertThat(e1.getPooled(), equalTo("held"));
+        assertThat(pool.size(), is(2));
+        assertThat(pool.getReservedCount(), is(0));
+        assertThat(pool.getIdleCount(), is(1));
+        assertThat(pool.getInUseCount(), is(1));
+
+        Pool.Entry<String> e2 = pool.reserve();
+        e2.enable("leaked", true);
+        assertThat(e2.getPooled(), equalTo("leaked"));
+        assertThat(pool.size(), is(3));
+        assertThat(pool.getReservedCount(), is(0));
+        assertThat(pool.getIdleCount(), is(1));
+        assertThat(pool.getInUseCount(), is(2));
+        e2 = null;
+        System.gc();
+        pool.sweep();
+
+        assertThat(pool.size(), is(2));
```

Comment:
This assertion is probably flaky as there is no guarantee that after `System.gc()` any actually leaked object will see its weak ref cleared. 

I think awaitility can help here:

```java
await().atMost(5, TimeUnit.SECONDS).until(() ->
{
    System.gc();
    pool.sweep();
    return pool.size();
}, is(2));
```